docs/hugo/content/design/ADR-2022-09-Multiple-Credential-Operator.md

```diff
@@ -0,0 +1,153 @@
+```
+annotations: 
+  serviceoperator.azure.com/credential-from: any-secret
+```
+
+**Pros:**
+1. Users don't have to specify the configuration for each resource explicitly if using a namespaced secret
+2. Users get the flexibility of over-riding the credential they want to use as per resource and as per resource group
+
+**Cons:**
+1. Could be complicated/confusing for users to understand the usage and credentials being used
+
+### Option 4: Secret lookup + configuration using global secret
+
+This option, relates to the option 3. However, the change is how we configure to use the per-resource-group and per-resource secret. We can have a unique slice property(CREDENTIAL_FROM) in the global secret where admin users can specify the secrets they want to load on the startup.
+Then only these loaded secrets would be used further for per-resource and per-resource-group. This option would restrict the other non-admin users(who don't have access to the global secret) to add their credential. 
+
+**Pros:**
+1. Users don't have to specify the configuration for each resource explicitly if using a namespaced secret
+2. Security conscious, not anyone in the cluster can add to use their credentials.
+
+**Cons:**
+1. Could be complicated/confusing for users to understand the usage and credentials being used
+2. Not very flexible, as pod will have to startup again if a new secret needs to be loaded
+
+## Handling for the failure cases and logging
+
+- Failure on loading a secret would result in failure on resource creation
+- Failure on authenticating the credentials would result in failure on resource creation
+- Logging should be clear about which credential is being used for a resource
+- Logging should be clear about the credential load/authentication failure
+
+## Milestones
+
+### Support for namespaced secrets
+
+We aim to do the initial work needed in the operator to support this type of multitenancy and add support for namespaced secrets.
+
+**Timelines:** TBD
+
+### Support for per-resource-group secrets
+
+We aim to support per-resource-group secrets under this milestone and configuration work required.
+
+**Timelines:** TBD
+
+### Support for per-resource secrets
+
+We aim to support per-resource secrets under this milestone and configuration work required.
+
+**Timelines:** TBD
+
+### Support for credential cache
+
+TBD
+
+## Open questions
+1. Do we need to support both per-resource and per-resource-group credentials?
+2. Do we want a serviceoperator.azure.com/subscription-id annotation that we write? We can use this to protect against users accidentally updating credential to move resources between subscriptions.
+3. Spell out what fields are in aso-credential. clientid/subscriptionid/tenantId/etc?
+4. How are we going to make this work for Managed Identity?
+5. Replace the above-mentioned(in options) annotations with spec property?
+
+## Consequences
+
+TBC
+
+## Experience Report
+
+TBC
+
+## References
+
+TBC
```

Comment:
Do we need to use a persistent cache like that? In conversations so far, I've assumed that any caching would be in-memory, to avoid hitting the api-server any more often than needed.

If the ASO pod restarts, loading all the secrets afresh (not from the cache) is a feature, as it gives users a simple (if draconian) way to ensure that up to date secrets are being used.